modules/partners/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+		/// Partner's metadata is updated
+		PartnerMetadataUpdated { partner_id: PartnerId },
+	}
+
+	#[pallet::storage]
+	pub type NextId<T: Config> = StorageValue<_, PartnerId, ValueQuery>;
+
+	#[pallet::storage]
+	pub type PartnerMetadata<T: Config> =
+		StorageMap<_, Identity, PartnerId, BoundedVec<u8, T::MaxMetadataLength>, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Referral<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, ReferralInfo<T::BlockNumber>, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn register_partner(
+			origin: OriginFor<T>,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// transfer registration fee to treasury
+			let treasury_account = T::TreasuryAccount::get();
+			<T as module::Config>::Currencies::transfer(
+				T::GetNativeCurrencyId::get(),
+				&who,
+				&treasury_account,
+				T::RegisterFee::get(),
+			)?;
+
+			Self::feeless_register_partner(who, metadata)
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn set_referral(origin: OriginFor<T>, partner: PartnerId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Referral::<T>::insert(
+				&who,
+				ReferralInfo {
+					partner_id: partner,
+					expiry: T::BlockNumberProvider::current_block_number().saturating_add(T::ReferralExpire::get()),
+				},
+			);
+
+			Self::deposit_event(Event::ReferralSet {
+				who,
+				partner_id: partner,
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn admin_register_partner(
+			origin: OriginFor<T>,
+			owner: T::AccountId,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			Self::feeless_register_partner(owner, metadata)
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn update_partner(
+			origin: OriginFor<T>,
+			partner: PartnerId,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			PartnerMetadata::<T>::insert(&partner, metadata);
+
+			Self::deposit_event(Event::PartnerMetadataUpdated { partner_id: partner });
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn feeless_register_partner(owner: T::AccountId, metadata: BoundedVec<u8, T::MaxMetadataLength>) -> DispatchResult {
+		let id = NextId::<T>::get();
+		// generate sub account, this shouldn't fail
+		let sub_account: T::AccountId = T::PalletId::get()
+			.try_into_sub_account(id)
+			.ok_or(Error::<T>::SubAccountGenerationFailed)?;
+
+		// increment NextId and add partner metadata to storage
+		NextId::<T>::put(id.checked_add(One::one()).ok_or(ArithmeticError::Overflow)?);
+		PartnerMetadata::<T>::insert(id, metadata);
+
+		// make caller full proxy (default is the most permissioned value) to the now registered sub account
+		// Still requires proxy fee
```

Comment:
Proxy requires fees to work, maybe transfer fee amount from treasury to account?